pkg/util/winutil/scmmonitor.go

```diff
@@ -0,0 +1,203 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+//go:build windows
+// +build windows
+
+package winutil
+
+import (
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+	"golang.org/x/sys/windows/svc/mgr"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type ServiceInfo struct {
+	ServiceName []string
+	DisplayName []string
+}
+
+// Service is the return value from a query by pid.
+type Service struct {
+	essp      []*windows.ENUM_SERVICE_STATUS_PROCESS
+	startTime uint64
+}
+
+// SCMMonitor is an object that allows the caller to monitor Windows services.
+// The object will maintain a table of active services indexed by PID
+type SCMMonitor struct {
+	mux              sync.Mutex
+	pidToService     map[uint64]*Service
+	nonServicePid    map[uint64]uint64 // start time of this pid
+	lastMapTime      uint64            // ns since jan1 1970
+	serviceRefreshes uint64
+}
+
+// GetServiceMonitor returns a service monitor object
+func GetServiceMonitor() *SCMMonitor {
+	return &SCMMonitor{
+		pidToService:  make(map[uint64]*Service),
+		nonServicePid: make(map[uint64]uint64),
+	}
+}
+
+type StartTimeFunc func(uint64) (uint64, error)
+
+var pGetProcessStartTimeAsNs = StartTimeFunc(getProcessStartTimeAsNs)
+
+// GetRefreshCount returns the number of times we've actually queried
+// the SCM database.  used for logging stats.
+func (scm *SCMMonitor) GetRefreshCount() uint64 {
+	return scm.serviceRefreshes
+}
+func (scm *SCMMonitor) refreshCache() error {
+
+	// EnumServiceStatusEx requires only SC_MANAGER_ENUM_SERVICE.  Switch to
+	// new library to use least privilege
+	h, err := windows.OpenSCManager(nil, nil, windows.SC_MANAGER_ENUMERATE_SERVICE)
+	if err != nil {
+		log.Warnf("Failed to connect to scm %v", err)
+		return fmt.Errorf("Failed to open SCM %v", err)
+	}
+	m := &mgr.Mgr{Handle: h}
+	defer m.Disconnect()
+
+	var bytesNeeded, servicesReturned uint32
+	var buf []byte
+	for {
+		var p *byte
+		if len(buf) > 0 {
+			p = &buf[0]
+		}
+		err = windows.EnumServicesStatusEx(m.Handle, windows.SC_ENUM_PROCESS_INFO,
```

Comment:
What does the performance look like for `refreshCache`? Would it be possible to add benchmarks?

I realize that we're only calling this a single time per process check run, but I don't have a good sense of how long these API calls take, and think it might be a bit helpful.